Feature/circleci and docker

olevs opened the pull request:
Initial CircleCI setup with custom Docker image for the ALive project

olevs committed:
Added Dockerfile for creating a Docker container containing a compiled Slicer, with all build files

olevs committed:
ENH: Added Dockerfile for creating container with precompiled Slicer + ALive plugin

olevs committed:
ENH: Use ALive fork of Slicer in Dockerfile

olevs committed:
ENH: Added initial circleci config file. Not finished

olevs committed:
ENH: Updated circleci config file

olevs committed:
ENH: Clone with https instead of git

olevs committed:
Merge branch 'feature/circleci_and_docker' of github.com:ALive-research/Slicer-LiverAnalysis into feature/circleci_and_docker

RafaelPalomar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,47 @@
+  build:
+    working_directory: /usr/src/Slicer-LiverAnalysis
+    docker:
+      - image: olevs/slicer-alive-build
```
We should use [slicer/slicer-build](https://hub.docker.com/r/slicer/slicer-build). It has all we need and it is maintained by the Slicer devs

olevs replied:
slicer/slicer-build don't work, and isn't maintained. It's slicer/slicer-base that is the image to use.
Using this image will add +1,5 hours to each build as the build files are deleted. It don't seem like Slicer keeps any docker images with build files. That's the main reason for creating a new image.

olevs replied:
To get something up and running I used our fork of Slicer, as the latest changes in the main fork haven't compiled for the last week. I just tried again to compile it inside slicer/slicer-build, and it still fails.

RafaelPalomar replied:
@olevs  you are right, the slicer-build image looks not maintained. Then I suggest the following:

1. Keep all the needed docker image definitions for ALive in a single repository ([ALive-Docker](https://github.com/ALive-research/ALive-Docker)). For now, we can create our own `slicer-build` based on `SlicerDocker/slicer-base`. I would remove  the `slicer-alive-build` in favor of checkout/builld/test commands in the circleci configuration. It is easier to maintain.
2. Have the container `slicer-build` image in [aliveresearch/slicer-build](https://hub.docker.com/repository/docker/aliveresearch/slicer-build) dockerhub repo.

olevs replied:
I agree. This sounds like a good solution. All the ALive stuff should be lightweight anyway, and don't need to be present in the Docker container.

RafaelPalomar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,47 @@
+      - run:
+          name: Update 3D Slicer
+          command: |
+            cd /usr/src/Slicer
+            git checkout .
+            git pull
```
If we use the [slicer/slicer-build](https://hub.docker.com/r/slicer/slicer-build) image, here we would do the git clone, etc..

olevs replied:
Using aliveresearch/slicer-build seems to work

RafaelPalomar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,47 @@
+      - run:
+          name: Build LiverAnalysis plugin
+          environment:
+            BUILD_TOOL_FLAGS: "-j3"
```
I would leave this as `-j1` just to start. I think extensions will not take much to compile and we will probably get only one core with the free plan anyways

olevs replied:
I set it to -j1 for now. It probably don't matter much. Slicer uses -j3.

RafaelPalomar commented on .circleci/config.yml:
```diff
@@ -0,0 +1,47 @@
+      #      git pull
+      
+      - run:
+          name: Build LiverAnalysis plugin
```
plugin -> extension. In the slicer jargon there are no plugins, but  extensions or modules. Let's try to keep their language (sorry for the pickiness :)

olevs replied:
Changed

RafaelPalomar committed:
Delete Dockerfile

I think this file is not necessary. The idea would be to use the [slicer/slicer-build](https://hub.docker.com/r/slicer/slicer-build) docker image as a base and move the rest of the commands to the circleci configuration

RafaelPalomar committed:
Delete Dockerfile

I think this is not necessary. We can use the [slicer/slicer-build](https://hub.docker.com/r/slicer/slicer-build) docker image

RafaelPalomar reviewed (changes requested):
Thanks for working on this, @olevs! 

Long story short: I think we should simplify the CI system to only use the [slicer/slicer-build](https://hub.docker.com/r/slicer/slicer-build) docker image as a base. Thus, we don't have to maintain any docker images. Note that we don't wan to diverge from the official Slicer (therefore no need to use our fork of Slicer).

olevs committed:
ENH: Updated CircleCI config file to use aliveresearch/slicer-build Docker image

olevs committed:
COMP: Fixed bug in CircleCI config

RafaelPalomar commented:
We will need to add a testing section in the CircleCI configuration. But for now, since we don't have tests, we can approve this.

RafaelPalomar commented:
Great work @olevs!
